File: lua/day4_part2_2018.lua
-- Read input from file
local file = io.open("input.txt", "r")
local input = file:read("*a")
file:close()

-- Parse input
local entries = {}
for line in input:gmatch("[^\n]+") do
    local timestamp, event = line:match("%[([^%]]+)%] (.+)")
    local year, month, day, hour, minute = timestamp:match("(%d+)-(%d+)-(%d+) (%d+):(%d+)")
    local guard_id = event:match("#(%d+)")
    local action = event:match("falls asleep") and "falls_asleep" or event:match("wakes up") and "wakes_up" or "begins_shift"
    table.insert(entries, {
        year = tonumber(year),
        month = tonumber(month),
        day = tonumber(day),
        hour = tonumber(hour),
        minute = tonumber(minute),
        guard_id = tonumber(guard_id),
        action = action
    })
end

-- Sort entries by timestamp
table.sort(entries, function(a, b)
    if a.year ~= b.year then return a.year < b.year end
    if a.month ~= b.month then return a.month < b.month end
    if a.day ~= b.day then return a.day < b.day end
    if a.hour ~= b.hour then return a.hour < b.hour end
    return a.minute < b.minute
end)

-- Calculate total sleep time for each guard
local guard_sleep_times = {}
local current_guard = nil
local sleep_start = nil
for _, entry in ipairs(entries) do
    if entry.action == "begins_shift" then
        current_guard = entry.guard_id
        guard_sleep_times[current_guard] = guard_sleep_times[current_guard] or 0
    elseif entry.action == "falls_asleep" then
        sleep_start = entry.minute
    elseif entry.action == "wakes_up" then
        guard_sleep_times[current_guard] = guard_sleep_times[current_guard] + (entry.minute - sleep_start)
    end
end

-- Find the guard with the most total sleep time
local max_sleep_time = 0
local max_sleep_guard = nil
for guard_id, sleep_time in pairs(guard_sleep_times) do
    if sleep_time > max_sleep_time then
        max_sleep_time = sleep_time
        max_sleep_guard = guard_id
    end
end

-- Find the minute the max sleep guard was asleep the most
local minute_counts = {}
current_guard = nil
sleep_start = nil
for _, entry in ipairs(entries) do
    if entry.action == "begins_shift" then
        current_guard = entry.guard_id
    elseif entry.action == "falls_asleep" then
        sleep_start = entry.minute
    elseif entry.action == "wakes_up" then
        for i = sleep_start, entry.minute - 1 do
            minute_counts[i] = (minute_counts[i] or 0) + (current_guard == max_sleep_guard and 1 or 0)
        end
    end
end

local max_minute = 0
local max_minute_count = 0
for minute, count in pairs(minute_counts) do
    if count > max_minute_count then
        max_minute = minute
        max_minute_count = count
    end
end

-- Part 1 answer
print(max_sleep_guard * max_minute)

-- Part 2 answer
local max_minute_guard = nil
local max_minute_count = 0
for guard_id, sleep_times in pairs(guard_sleep_times) do
    minute_counts = {}
    current_guard = nil
    sleep_start = nil
    for _, entry in ipairs(entries) do
        if entry.action == "begins_shift" then
            current_guard = entry.guard_id
        elseif entry.action == "falls_asleep" then
            sleep_start = entry.minute
        elseif entry.action == "wakes_up" then
            for i = sleep_start, entry.minute - 1 do
                minute_counts[i] = (minute_counts[i] or 0) + (current_guard == guard_id and 1 or 0)
            end
        end
    end
    for minute, count in pairs(minute_counts) do
        if count > max_minute_count then
            max_minute_guard = guard_id
            max_minute = minute
            max_minute_count = count
        end
    end
end

print(max_minute_guard * max_minute)

File: rust/day4_part2_2018.rs

use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

#[derive(Debug)]
enum EventType {
    BeginsShift(u32),
    FallsAsleep,
    WakesUp,
}

#[derive(Debug)]
struct Event {
    year: u32,
    month: u32,
    day: u32,
    hour: u32,
    minute: u32,
    event_type: EventType,
}

fn parse_event(line: &str) -> Event {
    let parts: Vec<&str> = line.split(|c| c == '[' || c == ']').collect();
    let date_time: Vec<&str> = parts[1].split(' ').collect();
    let date: Vec<&str> = date_time[0].split('-').collect();
    let time: Vec<&str> = date_time[1].split(':').collect();
    let event_str = parts[2].trim();

    let year = date[0].parse().unwrap();
    let month = date[1].parse().unwrap();
    let day = date[2].parse().unwrap();
    let hour = time[0].parse().unwrap();
    let minute = time[1].parse().unwrap();

    let event_type = if event_str.starts_with("Guard") {
        let guard_id: String = event_str
            .split_whitespace()
            .nth(1)
            .unwrap()
            .chars()
            .skip(1)
            .collect();
        EventType::BeginsShift(guard_id.parse().unwrap())
    } else if event_str == "falls asleep" {
        EventType::FallsAsleep
    } else {
        EventType::WakesUp
    };

    Event {
        year,
        month,
        day,
        hour,
        minute,
        event_type,
    }
}

fn main() {
    let file = File::open("input.txt").expect("Failed to open input file");
    let reader = BufReader::new(file);

    let mut events: Vec<Event> = reader
        .lines()
        .map(|line| parse_event(&line.unwrap()))
        .collect();

    events.sort_by(|a, b| {
        (a.year, a.month, a.day, a.hour, a.minute)
            .cmp(&(b.year, b.month, b.day, b.hour, b.minute))
    });

    let mut guard_sleep_times: HashMap<u32, Vec<u32>> = HashMap::new();
    let mut current_guard_id = 0;
    let mut sleep_start_minute = 0;

    for event in events {
        match event.event_type {
            EventType::BeginsShift(id) => current_guard_id = id,
            EventType::FallsAsleep => sleep_start_minute = event.minute,
            EventType::WakesUp => {
                let sleep_end_minute = event.minute;
                let sleep_times = guard_sleep_times.entry(current_guard_id).or_insert(vec![0; 60]);
                for minute in sleep_start_minute..sleep_end_minute {
                    sleep_times[minute as usize] += 1;
                }
            }
        }
    }

    // Strategy 1
    let (guard_id_most_sleep, sleep_minutes) = guard_sleep_times
        .iter()
        .max_by_key(|(_, minutes)| minutes.iter().sum::<u32>())
        .unwrap();

    let most_sleep_minute = sleep_minutes
        .iter()
        .enumerate()
        .max_by_key(|(_, &count)| count)
        .unwrap()
        .0;

    println!(
        "Strategy 1: Guard ID * Minute = {}",
        guard_id_most_sleep * (most_sleep_minute as u32)
    );

    // Strategy 2
    let (guard_id_most_frequent, minute_most_frequent) = guard_sleep_times
        .iter()
        .map(|(&id, minutes)| {
            let (minute, _) = minutes
                .iter()
                .enumerate()
                .max_by_key(|(_, &count)| count)
                .unwrap();
            (id, minute, minutes[minute])
        })
        .max_by_key(|&(_, _, count)| count)
        .map(|(id, minute, _)| (id, minute))
        .unwrap();

    println!(
        "Strategy 2: Guard ID * Minute = {}",
        guard_id_most_frequent * (minute_most_frequent as u32)
    );
}
